chain/consensus/common/crossmsg.go

```diff
@@ -0,0 +1,180 @@
+package common
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/lotus/blockstore"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/actors/builtin"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical/actors/sca"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical/subnet"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/vm"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+	blockadt "github.com/filecoin-project/specs-actors/actors/util/adt"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"golang.org/x/xerrors"
+)
+
+func checkCrossMsg(pstore, snstore blockadt.Store, parentSCA, snSCA *sca.SCAState, msg *types.Message) error {
+	switch hierarchical.GetMsgType(msg) {
+	case hierarchical.Fund:
+		// sanity-check: the root chain doesn't support topDown messages,
+		// so return an error if parentSCA is nil and we are here.
+		if parentSCA == nil {
+			return xerrors.Errorf("root chains (id=%v) does not support topDown cross msgs", snSCA.NetworkName)
+		}
+		return checkTopDownMsg(pstore, parentSCA, snSCA, msg)
+	case hierarchical.Release:
+		panic("Not implemented")
+	case hierarchical.Cross:
+		panic("Not implemented")
+	}
+
+	return xerrors.Errorf("Unknown cross-msg type")
+}
+
+// checkTopDownMsg validates the topdown message.
+// - It checks that the msg nonce is larger than AppliedDownTopNonce in the subnet SCA
+// Recall that applying crossMessages increases the AppliedNonce of the SCA in the subnet
+// where the message is applied.
+// - It checks that the cross-msg is committed in the sca of the parent chain
+func checkTopDownMsg(pstore blockadt.Store, parentSCA, snSCA *sca.SCAState, msg *types.Message) error {
```

Comment:
who calls this? shouldn't it only be called if the messages is committed in the parent? why do we need to recheck in L53